### PR TITLE
BUG: Fix strange behavior of infinite-step-size/underflow-case in arange

### LINKED ISSUE #10206: BUG: Strange behavior of infinite step size in arange
```python
>>> np.arange(0, 1, 1e10)
array([0.])
>>> np.arange(0, 1, 1e100)
array([0.])
>>> np.arange(0, 1, 1e1000)  # overflows
array([], dtype=float64)
>>> np.arange(0, 1, np.inf)
array([], dtype=float64)
```

I think the last two cases should match the others.

### PR DESCRIPTION
**Pull request by Licht-T**
This closes https://github.com/numpy/numpy/issues/10206.

The problem is
```
>>> np.arange(0, 1, 1e10)
array([0.])
>>> np.arange(0, 1, 1e100)
array([0.])
>>> np.arange(0, 1, 1e1000)  # overflows
array([], dtype=float64)
>>> np.arange(0, 1, np.inf)
array([], dtype=float64)
```
Where the step size becomes infinite.

### CODE REVIEW

File: numpy/core/src/multiarray/ctors.c [hunk 1]
```diff
@@ -3083,11 +3106,23 @@ _calc_length(PyObject *start, PyObject *stop, PyObject *step, PyObject **next, i
         if (error_converting(value)) {
             return -1;
         }
-        len = _arange_safe_ceil_to_intp(value);
-        if (error_converting(len)) {
-            return -1;
+
+        if (val_is_zero && next_is_nonzero) {
+            if (signbit(value)) {
```
Comment (xoviat) [inline numpy/core/src/multiarray/ctors.c]: Use `npy_signbit`.

**Comment by Licht-T** [comment]
Thanks @xoviat, fixed!

File: numpy/core/src/multiarray/ctors.c [hunk 2]
```diff
@@ -2968,11 +2968,25 @@ PyArray_Arange(double start, double stop, double step, int type_num)
...
+    double delta, tmp_len;
     NPY_BEGIN_THREADS_DEF;
 
-    length = _arange_safe_ceil_to_intp((stop - start)/step);
-    if (error_converting(length)) {
-        return NULL;
+    delta = stop - start;
+    tmp_len = delta/step;
+
+    if (tmp_len == 0.0 && delta != 0.0) {
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/ctors.c]: This needs a comment explaining why it is a special case

Isn't this only satisfied if `step` is `inf` anyway?

Comment (Licht-T) [reply numpy/core/src/multiarray/ctors.c]: @eric-wieser This considers som underflow cases.
```python
>>> m = np.float32(3.40282347E+38)
>>> e = np.float32(1.401298E-45)
>>> m
3.4028235e+38
>>> e
1.4012985e-45
>>> e/m
0.0
>>> np.arange(np.float32(0), e, m)
array([], dtype=float64)
```

**Comment by Licht-T** [comment]
@eric-wieser Added comments.

File: numpy/core/src/multiarray/ctors.c [hunk 3]
```diff
@@ -2968,11 +2968,26 @@ PyArray_Arange(double start, double stop, double step, int type_num)
...
+    double delta, tmp_len;
     NPY_BEGIN_THREADS_DEF;
 
-    length = _arange_safe_ceil_to_intp((stop - start)/step);
-    if (error_converting(length)) {
-        return NULL;
+    delta = stop - start;
+    tmp_len = delta/step;
+
+    // Underflow and divide-by-inf check
```
Comment (charris) [inline numpy/core/src/multiarray/ctors.c]: Don't use C++ comments, use `/* ... */`.

Comment (Licht-T) [reply numpy/core/src/multiarray/ctors.c]: Thanks @charris, fixed!

File: numpy/core/src/multiarray/ctors.c [hunk 4]
```diff
@@ -3049,12 +3065,19 @@ _calc_length(PyObject *start, PyObject *stop, PyObject *step, PyObject **next, i
...
+    next_is_nonzero = PyObject_RichCompareBool(*next, zero, Py_NE);
     val = PyNumber_TrueDivide(*next, step);
     Py_DECREF(*next);
     *next = NULL;
+
     if (!val) {
         return -1;
     }
+
+    val_is_zero = PyObject_RichCompareBool(val, zero, Py_EQ);
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/ctors.c]: Should be an `int`, not an `intp`.

File: numpy/core/src/multiarray/ctors.c [hunk 5]
```diff
@@ -3049,12 +3065,19 @@ _calc_length(PyObject *start, PyObject *stop, PyObject *step, PyObject **next, i
         }
         return -1;
     }
+
+    next_is_nonzero = PyObject_RichCompareBool(*next, zero, Py_NE);
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/ctors.c]: Could use `PyObject_IsTrue`, but that might be worse on unusual types

File: numpy/core/src/multiarray/ctors.c [hunk 6]
```diff
@@ -3049,12 +3065,19 @@ _calc_length(PyObject *start, PyObject *stop, PyObject *step, PyObject **next, i
...
     val = PyNumber_TrueDivide(*next, step);
     Py_DECREF(*next);
     *next = NULL;
+
     if (!val) {
         return -1;
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/ctors.c]: `zero` is leaked here

File: numpy/core/src/multiarray/ctors.c (hunk 5 above)
Comment (Licht-T) [reply numpy/core/src/multiarray/ctors.c]: @eric-wieser IMO, `PyObject_RichCompareBool` is better because this is clear what we do.

**Comment by Licht-T** [comment]
Thanks @eric-wieser, fixed!

File: numpy/core/src/multiarray/ctors.c [hunk 7]
```diff
@@ -3049,12 +3066,20 @@ _calc_length(PyObject *start, PyObject *stop, PyObject *step, PyObject **next, i
         }
         return -1;
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/ctors.c]: Need to `DECREF` `zero` here too - or better yet, just allocate it further down.

File: numpy/core/src/multiarray/ctors.c [hunk 8]
```diff
@@ -3036,7 +3051,9 @@ static npy_intp
 _calc_length(PyObject *start, PyObject *stop, PyObject *step, PyObject **next, int cmplx)
 {
     npy_intp len, tmp;
+    PyObject *zero = PyInt_FromLong(0);
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/ctors.c]: Really you should check the return value is not `NULL` here too.

**Comment by Licht-T** [comment]
@eric-wieser Fixed!

**Comment by Licht-T** [comment]
Also, I need to add some underflow-case tests.

File: numpy/core/src/multiarray/ctors.c [hunk 9]
```diff
@@ -3036,11 +3051,19 @@ static npy_intp
...
     double value;
 
+    zero = PyInt_FromLong(0);
+    if (!zero) {
+        return -1;
+    }
+
     *next = PyNumber_Subtract(stop, start);
     if (!(*next)) {
+        Py_DECREF(zero);
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/ctors.c]: If you move the allocation of `zero` down below this `if`, then you don't need to clear it up here

**Comment by eric-wieser** [comment]
Did you mean to refer to  #10271 in the PR description? Seems unrelated to me

File: numpy/core/src/multiarray/ctors.c (hunk 9 above)
Comment (Licht-T) [reply numpy/core/src/multiarray/ctors.c]: @eric-wieser If I move down `zero` allocation, we need to `DECREF` `*next` and set `NULL` to `next` when the `zero` allocation failed. IMO, this is complicated.

**Comment by Licht-T** [comment]
@eric-wieser I did not refer to that issue. Maybe @charris added this.
> Did you mean to refer to #10271 in the PR description? Seems unrelated to me

File: numpy/core/src/multiarray/ctors.c (hunk 9 above)
Comment (eric-wieser) [reply numpy/core/src/multiarray/ctors.c]: Fair

**Comment by Licht-T** [comment]
@eric-wieser CI failed because of the strange compile warning.
Do you have any idea?
```
compile options: '-DNPY_INTERNAL_BUILD=1 -DHAVE_NPY_CONFIG_H=1 -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE=1 -D_LARGEFILE64_SOURCE=1 -DHAVE_CBLAS -Inumpy/core/src/private -Inumpy/core/include -Inumpy/core/include/numpy -Inumpy/core/src/private -Inumpy/core/src -Inumpy/core -Inumpy/core/src/npymath -Inumpy/core/src/multiarray -Inumpy/core/src/umath -Inumpy/core/src/npysort -I/home/rito/miniconda3/include/python3.6m -Inumpy/core/src/private -Inumpy/core/src/npymath -Inumpy/core/src/private -Inumpy/core/src/npymath -Inumpy/core/src/private -Inumpy/core/src/npymath -c'
gcc: numpy/core/src/multiarray/ctors.c
In file included from /home/rito/miniconda3/include/python3.6m/pytime.h:6:0,
                 from /home/rito/miniconda3/include/python3.6m/Python.h:65,
                 from numpy/core/src/multiarray/ctors.c:2:
numpy/core/src/multiarray/ctors.c: In function ‘PyArray_ArangeObj’:
/home/rito/miniconda3/include/python3.6m/object.h:790:12: warning: ‘next’ may be used uninitialized in this function [-Wmaybe-uninitialized]
         if (_Py_DEC_REFTOTAL  _Py_REF_DEBUG_COMMA       \
            ^
numpy/core/src/multiarray/ctors.c:3156:15: note: ‘next’ was declared here
     PyObject *next, *err;
               ^~~~
```

**Comment by Licht-T** [comment]
Ah..., seems that the macro of CPython includes `next`?

File: numpy/core/src/multiarray/ctors.c [hunk 10]
```diff
@@ -3036,11 +3051,19 @@ static npy_intp
...
 {
     npy_intp len, tmp;
-    PyObject *val;
+    PyObject *zero, *val;
+    int next_is_nonzero, val_is_zero;
     double value;
 
+    zero = PyInt_FromLong(0);
+    if (!zero) {
+        return -1;
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/ctors.c]: I think the problem is that `*next` is not yet set here.

Comment (eric-wieser) [reply numpy/core/src/multiarray/ctors.c]: Also, that `Py_DECREF(next);` should be `Py_XDECREF`...

File: numpy/core/src/multiarray/ctors.c (hunk 9 above)
Comment (eric-wieser) [reply numpy/core/src/multiarray/ctors.c]: >  and set NULL to *next when the zero allocation failed. 

You already need to do this

**Comment by Licht-T** [comment]
Thanks @eric-wieser, fixed!

**Comment by Licht-T** [comment]
Added the test for some underflow cases

File: numpy/core/src/multiarray/ctors.c [hunk 11]
```diff
@@ -3049,12 +3065,27 @@ _calc_length(PyObject *start, PyObject *stop, PyObject *step, PyObject **next, i
...
+
     if (!val) {
+        Py_DECREF(zero);
         return -1;
     }
+
+    val_is_zero = PyObject_RichCompareBool(val, zero, Py_EQ);
+    Py_DECREF(zero);
+
     if (cmplx && PyComplex_Check(val)) {
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/ctors.c]: Trying to work out if this section needs a fix too, but I can't work out what this is even trying to do.

Comment (Licht-T) [reply numpy/core/src/multiarray/ctors.c]: @eric-wieser Me too. I don't know why the complex case exists. This is none-sense as math.

**Comment by eric-wieser** [comment]
Can you extract `double _ceildiv(double, double)` and `PyObject * _ceildiv_O(PyObject *, PyObject *)` helper functions here, that handles all these corner cases for us?

File: numpy/core/src/multiarray/ctors.c (hunk 11 above)
Comment (eric-wieser) [reply numpy/core/src/multiarray/ctors.c]: See #10332.

**Review by eric-wieser — DISMISSED** [review]
On second thoughts, this is fine as is - I'll leave it open for a few more days to let someone else double-check refcounts, and then squash and merge your commits.

File: numpy/core/src/multiarray/ctors.c [hunk 12]
```diff
@@ -2968,11 +2968,26 @@ PyArray_Arange(double start, double stop, double step, int type_num)
...
+    delta = stop - start;
+    tmp_len = delta/step;
+
+    /* Underflow and divide-by-inf check */
+    if (tmp_len == 0.0 && delta != 0.0) {
+        if (npy_signbit(tmp_len)) {
+            length = -1.0;
+        }
+        else {
+            length = 1.0;
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/ctors.c]: nit: should be `1`, as this is an integer, not double, variable. Same above.

File: numpy/core/src/multiarray/ctors.c [hunk 13]
```diff
@@ -3083,11 +3114,24 @@ _calc_length(PyObject *start, PyObject *stop, PyObject *step, PyObject **next, i
...
-        if (error_converting(len)) {
-            return -1;
+
+        /* Underflow and divide-by-inf check */
+        if (val_is_zero && next_is_nonzero) {
+            if (npy_signbit(value)) {
+                len = -1.0;
+            }
+            else {
+                len = 1.0;
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/ctors.c]: Same here for int vs double

**Comment by Licht-T** [comment]
@eric-wieser This completely slipped my mind! I'm back, and it will be fixed in a couple of days.

**Comment by Licht-T** [comment]
@eric-wieser Fixed, and now all green!

File: numpy/core/src/multiarray/ctors.c [hunk 14]
```diff
@@ -3080,12 +3096,27 @@ _calc_length(PyObject *start, PyObject *stop, PyObject *step, PyObject **next, i
...
     val = PyNumber_TrueDivide(*next, step);
     Py_DECREF(*next);
     *next = NULL;
+
     if (!val) {
+        Py_DECREF(zero);
         return -1;
     }
+
+    val_is_zero = PyObject_RichCompareBool(val, zero, Py_EQ);
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/ctors.c]: You need to check that this doesn't return -1. Same above.

**Comment by Licht-T** [comment]
Thanks @eric-wieser, fixed!

**Review by eric-wieser — APPROVED** [review]

**Comment by eric-wieser** [comment]
Minor style nits, but easiest to fix them myself. Thanks for the perseverance - I'll leave this for a couple days to see if there are any more comments

**Comment by charris** [comment]
@eric-wieser  Just a ping.

**Comment by eric-wieser** [comment]
Thanks for the perseverance, @Licht-T - this will be in 1.15

**Pull request merged**